### DISCUSSION TITLE
chatgpt4-o integration

### DISCUSSION THREAD

**Discussion by ashepp** [judahpaul16/gpt-home#28] [Ideas]
Wondering whether it's an easy as a variable change to utilize ChatGPT4-o that was announced today? I'm wondering if it might help with some of the latency issues between the wake word and the response? Thanks again this is a fun project!

**Comment by judahpaul16** [comment]
When a model becomes available via the API it will become available here automatically if it includes `gpt`. 

![image](https://github.com/judahpaul16/gpt-home/assets/23343173/96071553-e4ef-43da-b7e9-2d6193d588c1)

See in `backend.py`:

```python
@app.post("/availableModels")
async def available_models():
    try:
        # Get available models from OpenAI
        openai.api_key = os.getenv("OPENAI_API_KEY")
        model_list = openai.Model.list()
        
        # Filter to only keep supported models.
        supported_models = [model['id'] for model in model_list['data'] if "gpt" in model['id'].lower()]

        return JSONResponse(content={"models": supported_models})
    except Exception as e:
        return HTTPException(status_code=500, detail=f"An error occurred: {str(e)}")
```

The latency is likely cause by the separate calls for speech-to-text and querying openai in the listen logic in `functions.py`. Currently each successful loop consists of recognizing what was said via the Google STT [speech recognition library](https://github.com/Uberi/speech_recognition), getting that response, parsing the response to see what is being asked and route the response to openai which then returns the final response to be spoken. Could possibly speed up latency by using a local recognizer but other ones I've tried are not as accurate. If your internet is fast the latency is not very noticeable in my testing.
